Specify Flow and Sender attributes and related Receiver parameter constraints (#12)
AMWA-TV/bcp-006-01: docs/NMOS With JPEG XS.md

### Sources

The Source resource MUST indicate `urn:x-nmos:format:video` for the `format`.
Source resources can be associated with many Flows at the same time.
The Source is therefore unaffected by the use of JPEG XS compression.

### Flows

The Flow resource MUST indicate `video/jxsv` in the `media_type` attribute, and `urn:x-nmos:format:video` for the `format`.
This has been permitted since IS-04 v1.1.

For Nodes implementing IS-04 v1.3 or higher, the following additional requirements on the Flow resource apply.

In addition to those attributes defined in IS-04 for all coded video Flows, the following attributes defined in the [Flow Attributes register](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/) of the [NMOS Parameter Registers][] are used for JPEG XS.

- [Components](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#components)  
  The Flow resource MUST indicate the color (sub-)sampling using the `components` attribute.
  The `components` array value corresponds to the `sampling`, `width` and `height` values in the SDP format-specific parameters defined by RFC 9134.

> - [Profile](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#profile)  
>   The Flow resource MUST indicate the JPEG XS profile, which defines limits on the required algorithmic features and parameter ranges used in the codestream.
>   The permitted `profile` values are strings, defined as per RFC 9134.
> - [Level](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#level)  
>   The Flow resource MUST indicate the JPEG XS level, which defines a lower bound on the required throughput for a decoder in the image (or decoded) domain.
>   The permitted `level` values are strings, defined as per RFC 9134.

Author: @tbr Would this be better? Or did we decide to reject describing Unrestricted profile/level?

```suggestion
- [Profile](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#profile)  
  The JPEG XS profile defines limits on the required algorithmic features and parameter ranges used in the codestream.
  When the `profile` attribute is omitted, the Unrestricted profile is implied.
  Otherwise, the `profile` value MUST be a string, defined as per RFC 9134, such as `High444.12`.
- [Level](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#level)  
  The JPEG XS level defines a lower bound on the required throughput for a decoder in the image (or decoded) domain.
  When the `level` attribute is omitted, the Unrestricted level is implied.
  Otherwise, the `level` values MUST be a string, defined as per RFC 9134, such as `1k-1`.
```

Reviewer: I guess you could write: When the `profile` attribute is omitted, the profile is unknown and shall be specified by the codestream.

This is how the RFC and SDP also interpret it.

> - [Profile](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#profile)  
>   The Flow resource MUST indicate the JPEG XS profile, which defines limits on the required algorithmic features and parameter ranges used in the codestream.
>   The permitted `profile` values are strings, defined as per RFC 9134.
> - [Level](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#level)  
>   The Flow resource MUST indicate the JPEG XS level, which defines a lower bound on the required throughput for a decoder in the image (or decoded) domain.
>   The permitted `level` values are strings, defined as per RFC 9134.

Author: @tbr Would this be better? Or did we decide to reject describing Unrestricted profile/level?

```suggestion
- [Profile](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#profile)  
  The JPEG XS profile defines limits on the required algorithmic features and parameter ranges used in the codestream.
  When the `profile` attribute is omitted, the Unrestricted profile is implied.
  Otherwise, the `profile` value MUST be a string, defined as per RFC 9134, such as `High444.12`.
- [Level](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#level)  
  The JPEG XS level defines a lower bound on the required throughput for a decoder in the image (or decoded) domain.
  When the `level` attribute is omitted, the Unrestricted level is implied.
  Otherwise, the `level` values MUST be a string, defined as per RFC 9134, such as `1k-1`.
```

Reviewer: I guess you could write: When the `level` attribute is omitted, the level is unknown and shall be specified by the codestream.

This is how the RFC and SDP also interpret it.

Author: Although then we don't have an explicit way to signal Unrestricted...

Reviewer: Not explicitly no... but one could say that omitting is equally difficult to signaling `Unrestricted` which not many implementations will handle (such decoder would need to start decoding and hope for the best, so I don;t see this as a practical approach in the field). At least if it is omitted, the stream still indicates it. If it's a well known profile/level/sublevel, the decoder will know if it can handle it or not.

- [Sublevel Bits Per Pixel](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#sublevel-bits-per-pixel)  
  The JPEG XS sublevel defines a lower bound on the required throughput for a decoder in the codestream (or coded) domain.
  When the `sublevel_bpp` attribute is omitted, the Full (or Unrestricted) sublevel is implied.
  Otherwise, the `sublevel_bpp` value SHOULD indicate the compression ratio used by the encoder in bits per pixel as a number.

>   When the precise value is not available, the value MUST be set based on the sublevel indicated in the codestream or in the SDP format-specific parameters defined by RFC 9134.

Reviewer: What does this mean?

Author: I think our intention is that if you encoded at 2.5 bpp, you _should_ put `"sublevel_bpp": 2.5`, but if that information isn't available at the NMOS layer you _must_ at least take the sublevel that willl be indicated in the codestream (and hopefully SDP data) such as `Sublev4bpp` and put in the corresponding values here, like `"sublevel_bpp": 4`.

Does that make sense? And if so, how should the explanation be improved?

Reviewer: That doesn't make sense to me. The parameter is called: `sublevel_bpp`, and I interpret this as a value that matches an XS sublevel. If you want to signal the bpp used for encoding, then the parameter should be just called `bpp` (or something). The sublevels are well-defined and can even be listed.

There is already the flow bit_rate, so actual bpp is redundant. Sublevel however is important for licensing or implementation restrictions etc.

Author: OK, if it's just representing the sublevel, I'm not entirely sure why we selected to use a number rather than enumerated strings like profile and level. I also thought we were accounting for implementations that were constrained more tightly, e.g. 1.5-4bpp.

Author: I understand the point about Flow bpp attribute being equivalently informative as Flow bit_rate, but since Receiver parameter constraints are in terms of individual attributes, the two constraints are not equivalently expressive, because one depends on frame rate, frame width and height.

Reviewer: The reason was that sublevels are inclusive and that as integers they can be specified via ranges or min max. When it's a string enum they would need to be all explicitly added in the negotiations.

>   When the precise value is not available, the value MUST be set based on the sublevel indicated in the codestream or in the SDP format-specific parameters defined by RFC 9134.

Reviewer 2: "When the precise value is not available, the value MUST be set based on the sublevel indicated in the codestream or in the SDP format-specific parameters defined by RFC 9134. For example, Sublev6bpp is indicated by the value 6"

I think it SHALL always correspond to the value encoded in the codestream which SHALL also correspond (if any) to the SDP transport file sublevel attribute.

Author: See https://github.com/AMWA-TV/bcp-006-01/pull/12#discussion_r891010937... or did we decide to not allow Flows to indicate the actual encoder target bpp (which may be lower than the sublevel indicated in the codestream and SDP)?

> - [Components](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#components)  
>   The Flow resource MUST indicate the color (sub-)sampling using the `components` attribute.
>   The `components` array value corresponds to the `sampling`, `width` and `height` values in the SDP format-specific parameters defined by RFC 9134.
> - [Profile](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#profile)  
>   The Flow resource MUST indicate the JPEG XS profile, which defines limits on the required algorithmic features and parameter ranges used in the codestream.
>   The permitted `profile` values are strings, defined as per RFC 9134.
> - [Level](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#level)  
>   The Flow resource MUST indicate the JPEG XS level, which defines a lower bound on the required throughput for a decoder in the image (or decoded) domain.
>   The permitted `level` values are strings, defined as per RFC 9134.
> - [Sublevel Bits Per Pixel](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#sublevel-bits-per-pixel)  
>   The JPEG XS sublevel defines a lower bound on the required throughput for a decoder in the codestream (or coded) domain.
>   When the `sublevel_bpp` attribute is omitted, the Full (or Unrestricted) sublevel is implied.
>   Otherwise, the `sublevel_bpp` value SHOULD indicate the compression ratio used by the encoder in bits per pixel as a number.
>   When the precise value is not available, the value MUST be set based on the sublevel indicated in the codestream or in the SDP format-specific parameters defined by RFC 9134.
>   For example, `Sublev6bpp` is indicated by the value `6`.
> - [Bit Rate](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#bit-rate)  
>   The Flow resource SHOULD indicate the target bit rate (kilobits/second) of the codestream.
>   The `bit_rate` integer value is expressed in units of 1000 bits per second, rounding up.

Author: ```suggestion
These attributes provide information for Controllers and Users to evaluate stream compatibility between Senders and Receivers.

- [Components](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#components)  
  The Flow resource MUST indicate the color (sub-)sampling using the `components` attribute.
  The `components` array value corresponds to the `sampling`, `width` and `height` values in the SDP format-specific parameters defined by RFC 9134.
- [Profile](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#profile)  
  The Flow resource is strongly RECOMMENDED to indicate the JPEG XS profile, which defines limits on the required algorithmic features and parameter ranges used in the codestream.
  The permitted `profile` values are strings, defined as per RFC 9134.
  The Unrestricted profile is indicated by omitting this attribute.
- [Level](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#level)  
  The Flow resource is strongly RECOMMENDED to indicate the JPEG XS level, which defines a lower bound on the required throughput for a decoder in the image (or decoded) domain.
  The permitted `level` values are strings, defined as per RFC 9134.
  The Unrestricted level is indicated by omitting this attribute.
- [Sublevel Bits Per Pixel](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#sublevel-bits-per-pixel)  
  The Flow resource is strongly RECOMMENDED to indicate the JPEG XS sublevel, which defines a lower bound on the required throughput for a decoder in the codestream (or coded) domain.
  The permitted `sublevel_bpp` values are integers, expressed in bits per pixel (BPP), corresponding to the sublevel.   
  The Full sublevel is indicated by the number of bits per component multiplied by the number of components in the uncompressed image.
  The Unrestricted sublevel is indicated by omitting this attribute.
- [Bit Rate](https://specs.amwa.tv/nmos-parameter-registers/branches/main/flow-attributes/#bit-rate)  
  The Flow resource SHOULD indicate the target bit rate (kilobits/second) of the codestream.
  The `bit_rate` integer value is expressed in units of 1000 bits per second, rounding up.
```

The SDP file at the `manifest_href` MUST comply with the requirements of RFC 9134.
If the Sender meets the traffic shaping and delivery timing requirements specified for ST 2110-22, the SDP file MUST also comply with the provisions of ST 2110-22.

For Nodes implementing IS-04 v1.3 or higher, the following additional requirements on the Sender resource apply.

In addition to those attributes defined in IS-04 for Senders, the following attributes defined in the [Sender Attributes register](https://specs.amwa.tv/nmos-parameter-registers/branches/main/sender-attributes/) of the NMOS Parameter Registers are used for JPEG XS.

- [Bit Rate](https://specs.amwa.tv/nmos-parameter-registers/branches/main/sender-attributes/#bit-rate)  
  If the Sender complies with ST 2110-22, the Sender resource MUST indicate the bit rate (kilobits/second) including the RTP transport overhead.
  Otherwise, it is only RECOMMENDED that this attribute is included.
  The `bit_rate` integer value is expressed in units of 1000 bits per second, rounding up.

>   The value is for the IP packets, including the RTP, UDP and IP packet headers and the payload.

Reviewer 2: Could we formulate such as to already support other transports like NDI and SRT ... We could refer to "The value is for IP packets encapsulated into their specific transport. For example for the RTP transport type it includes the RTP, UDP and IP packet headers and the payload.

Author: Yes, this is a good idea, so long as different specs don't want to define bit rate at higher or lower layers of encapsulation, e.g. Ethernet or UDP.

Author: ```suggestion
  The value is for the IP packets, so for the RTP payload format per RFC 9134, that includes the RTP, UDP and IP packet headers and the payload.
```

An example Sender resource is provided in the [Examples](../examples/).

## JPEG XS IS-04 Receivers

Nodes capable of receiving JPEG XS video streams MUST have a Receiver resource in the IS-04 Node API, which lists `video/jxsv` in the `media_types` array within the `caps` object.
This has been permitted since IS-04 v1.1.

Nodes implementing [BCP-004-01][] Receiver Capabilities use the existing `constraint_sets` parameter within the `caps` object, describing combinations of frame rates, width and height, and other parameters which the receiver can support, using the parameter constraints defined in the [Capabilities register](https://specs.amwa.tv/nmos-parameter-registers/branches/main/capabilities/) of the NMOS Parameter Registers.

Receivers are RECOMMENDED to use the following parameter constraints:

- [Profile](https://specs.amwa.tv/nmos-parameter-registers/branches/main/capabilities/#profile)
- [Level](https://specs.amwa.tv/nmos-parameter-registers/branches/main/capabilities/#level)
- [Sublevel Bits Per Pixel](https://specs.amwa.tv/nmos-parameter-registers/branches/main/capabilities/#sublevel-bits-per-pixel)  

>   When the JPEG XS decoder supports the Full (or Unrestricted) sublevel, the Receiver MAY indicate that this parameter is unconstrained, as per BCP-004-01.

Reviewer 2: Could the Full value be specified as the sum of the bit_depth of all the components? This would be the true Full value and should be allowed. Then unconstrained corresponds to not having or having an empty sublevel_bpp constraint.

Author: 👍 Let's address this alongside whether precise encoder target bpp is allowed.

Reviewer: Indeed, that's correct. In XS standard, `Full` translates to "bpc * num_components" (bpc is bits-per-component, which is signaled in the stream and is the same value for all components).

>   When the JPEG XS decoder supports the Full (or Unrestricted) sublevel, the Receiver MAY indicate that this parameter is unconstrained, as per BCP-004-01.

Author: ```suggestion

When the JPEG XS decoder supports the Unrestricted profile, level or sublevel, the Receiver MAY indicate that the parameter is unconstrained, as described in BCP-004-01.
```
